### PR TITLE
New features: alternative HTTP/S ports, fix download commercial license, alternative ssl directory

### PR DESCRIPTION
**Pull request by kristofvandam**
#### Pull Request (PR) description
- Allow to change default HTTP and HTTPS ports (handy when you use caddy behind a proxy)
- The commercial license option did not work due to a missing option to parse account id and api key
- Sane defaults for the systemd unit file (from https://github.com/mholt/caddy/blob/master/dist/init/linux-systemd/caddy.service) 
- Allow to choose where caddy stores it letsencrypt certificates (ssl_dir) (handy if you want a alternate directory like a shared store)

### CODE REVIEW

File: manifests/config.pp [hunk 1]
```diff
@@ -19,6 +19,14 @@
     managehome => true,
   }
 
+  file {$::caddy::caddy_ssl_dir:
```
Comment (bastelfreak) [inline manifests/config.pp]: Hi, please don't add `::` directly after the `$`. This is a deprecated style that was required for Puppet 3.x

File: manifests/package.pp [hunk 2]
```diff
@@ -9,12 +9,16 @@
...
   }
 
-  $caddy_url    = 'https://caddyserver.com/download/linux'
-  $caddy_dl_url = "${caddy_url}/${caddy::params::arch}?plugins=${caddy::caddy_features}&license=${caddy::caddy_license}&telemetry=${caddy::caddy_telemetry}"
-  $caddy_dl_dir = "${caddy::params::caddy_tmp_dir}/caddy_linux_${$caddy::params::arch}_custom.tar.gz"
+  $caddy_url        = 'https://caddyserver.com/download/linux'
+  $caddy_dl_url     = "${caddy_url}/${caddy::params::arch}?plugins=${caddy::caddy_features}&license=${caddy::caddy_license}&telemetry=${caddy::caddy_telemetry}"
+  $caddy_dl_dir     = "${caddy::params::caddy_tmp_dir}/caddy_linux_${$caddy::params::arch}_custom.tar.gz"
+  $caddy_dl_command = $caddy_license ? {
+    'personal'   => "curl -o ${caddy_dl_dir} \"${caddy_dl_url}\"",
```
Comment (bastelfreak) [inline manifests/package.pp]: Can't we use singlequotes within the string instead of escaped double quotes?

File: manifests/init.pp [hunk 3]
```diff
@@ -28,15 +28,20 @@
...
-  String                         $caddy_features  = 'http.filter,http.git,http.ipfilter'
+  String                         $install_path     = $caddy::params::install_path,
+  String                         $caddy_user       = $caddy::params::caddy_user,
+  String                         $caddy_group      = $caddy::params::caddy_group,
+  String                         $caddy_log_dir    = $caddy::params::caddy_log_dir,
+  String                         $caddy_tmp_dir    = $caddy::params::caddy_tmp_dir,
+  String                         $caddy_ssl_dir    = $caddy::params::caddy_ssl_dir,
+  String                         $caddy_home       = $caddy::params::caddy_home,
+  Enum['personal', 'commercial'] $caddy_license    = 'personal',
+  String                         $caddy_account_id = '',
```
Comment (bastelfreak) [inline manifests/init.pp]: Instead of empty string, can you default to undef?
```suggestion
   Optional[String[1]]           $caddy_account_id,
```

Same applies for `$caddy_api_key` We defined that some time ago in our [review guidelines](https://voxpupuli.org/docs/#reviewing-a-module-pr).

File: manifests/init.pp [hunk 4]
```diff
@@ -28,15 +28,20 @@
...
+  String                         $caddy_api_key    = '',
+  Enum['on','off']               $caddy_telemetry  = 'off',
+  String                         $caddy_features   = 'http.filter,http.git,http.ipfilter',
+  Integer                        $caddy_http_port  = $caddy::params::caddy_http_port,
```
Comment (bastelfreak) [inline manifests/init.pp]: There exists  a port datatype that can be used: https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/port.pp

File: templates/etc/systemd/system/caddy.service.erb [hunk 5]
```diff
@@ -9,19 +9,20 @@ Wants=network-online.target systemd-networkd-wait-online.service
...
+ExecStart=<%= @install_path %>/caddy -log <%= @caddy_log_dir %>/caddy.log -agree=true -conf=/etc/caddy/Caddyfile -root=/var/tmp -http-port=<%= @caddy_http_port %> -https-port=<%= @caddy_https_port %>
 ExecReload=/bin/kill -USR1 $MAINPID
 
 LimitNOFILE=1048576
-LimitNPROC=64
+LimitNPROC=512
 
 PrivateTmp=true
-PrivateDevices=true
+PrivateDevices=false
```
Comment (bastelfreak) [inline templates/etc/systemd/system/caddy.service.erb]: why was this changed?

File: templates/etc/systemd/system/caddy.service.erb [hunk 6]
```diff
@@ -9,19 +9,20 @@ Wants=network-online.target systemd-networkd-wait-online.service
...
+Environment=HOME=<%= @caddy_home %>
+Environment=CADDYPATH=<%= @caddy_ssl_dir %>
 
-ExecStart=<%= @install_path %>/caddy -log <%= @caddy_log_dir %>/caddy.log -agree=true -conf=/etc/caddy/Caddyfile -root=/var/tmp
+ExecStart=<%= @install_path %>/caddy -log <%= @caddy_log_dir %>/caddy.log -agree=true -conf=/etc/caddy/Caddyfile -root=/var/tmp -http-port=<%= @caddy_http_port %> -https-port=<%= @caddy_https_port %>
 ExecReload=/bin/kill -USR1 $MAINPID
 
 LimitNOFILE=1048576
-LimitNPROC=64
+LimitNPROC=512
```
Comment (bastelfreak) [inline templates/etc/systemd/system/caddy.service.erb]: Isn't 64 already quite high? Maybe this should be a parameter too?

**Comment by bastelfreak** [comment]
Hi @kristofvandam, thanks for the PR. I made a few inline comments, can you please have a look?

File: manifests/init.pp (hunk 4 above)
Comment (bastelfreak) [reply manifests/init.pp]: stdlib is missing as a dependency in `metadata.json`, please add it:

```json
  "dependencies": [
    {
      "name": "puppetlabs/stdlib",
      "version_requirement": ">= 4.25.0 < 6.0.0"
    }
  ]
```

File: templates/etc/systemd/system/caddy.service.erb (hunk 6 above)
Comment (kristofvandam) [reply templates/etc/systemd/system/caddy.service.erb]: I've changed this because it is recommended by the creator of Caddy, as he notes:

"Unmodified caddy is not expected to use more than that."

Which means it is possible that a default caddy configuration could use more than 64 processes, but less than 512. (according to mholt)

ref: https://github.com/mholt/caddy/blob/master/dist/init/linux-systemd/caddy.service

File: templates/etc/systemd/system/caddy.service.erb (hunk 5 above)
Comment (kristofvandam) [reply templates/etc/systemd/system/caddy.service.erb]: I've had some troubles with the original configuration, this change slipped in while troubleshooting. Though, the recommended systemd file also disables PrivateDevices as this does not work on some devices.

ref: https://github.com/mholt/caddy/commit/fd3fafa50caf0dcbe695d28b48198a1e2bf810bd

File: manifests/init.pp (hunk 4 above)
Comment (kristofvandam) [reply manifests/init.pp]: added, will soon commit

File: manifests/package.pp (hunk 2 above)
Comment (kristofvandam) [reply manifests/package.pp]: changed, will test and commit if ok

File: manifests/config.pp (hunk 1 above)
Comment (kristofvandam) [reply manifests/config.pp]: changed, also for previously defined variables.

File: manifests/config.pp [hunk 7]
```diff
@@ -69,15 +69,9 @@
         owner   => 'root',
         group   => 'root',
         content => template('caddy/etc/systemd/system/caddy.service.erb'),
-        notify  => Exec['systemctl-daemon-reload'],
+        notify  => Class['systemd::systemctl::daemon_reload'],
         require => Class['caddy::package'],
```
Comment (kristofvandam) [inline manifests/config.pp]: should i add a dependency for the systemd module?

Comment (kristofvandam) [reply manifests/config.pp]: undo done, are there any other remarks regarding this PR?

File: manifests/config.pp [hunk 8]
```diff
@@ -53,7 +61,7 @@
     require => User[$caddy::caddy_user],
   }
 
-  case $::operatingsystemmajrelease {
+  case $operatingsystemmajrelease {
```
Comment (bastelfreak) [inline manifests/config.pp]: this won't work, because it's a fact.Please undo the change.

File: manifests/config.pp [hunk 9]
```diff
@@ -64,12 +72,6 @@
         notify  => Exec['systemctl-daemon-reload'],
         require => Class['caddy::package'],
       }
-
-      exec {'systemctl-daemon-reload':
```
Comment (bastelfreak) [inline manifests/config.pp]: the exec also needs to come back, the file resource references it in line 72.

Comment (kristofvandam) [reply manifests/config.pp]: missed that, fixed.

File: manifests/config.pp (hunk 8 above)
Comment (kristofvandam) [reply manifests/config.pp]: done

**Comment by bastelfreak** [comment]
Hi,
I squashed this together in https://github.com/voxpupuli/puppet-caddy/pull/12 and will merge it after travis turns green. Thanks for your work!
